nlpreprcessing/annotation2BIO.py

```diff
@@ -0,0 +1,275 @@
+                        else:
+                            if offset_start == en_s:
+                                token.append("-".join(['B', en_type]))
+                                if offset_end >= en_e:
+                                    entity = next(entities_iter, None)
+                            else:
+                                logger.error(f"{token}\t{entity} not matched by their offsets.")
+                                token.append('O')
+                                entity = next(entities_iter, None)
+                    else:
+                        en_s = entity[2][0]
+                        en_e = entity[2][1]
+                        en_type = entity[1]
+                        if offset_start == en_s:
+                            token.append("-".join(['B', en_type]))
+                            if offset_end >= en_e:
+                                entity = next(entities_iter, None)
+                        elif offset_end < en_s:
+                            token.append('O')
+                        else:
+                            logger.error(f"{token}\t{entity} not matched by their offsets.")
+                            # token.append("-".join(['B', en_type]))
+                            token.append('O')
+                            entity = next(entities_iter, None)
+            nsent.append(token)
+        nsents.append(nsent)
+
+    sent_bound_range = dict()  # key: sent id; value: boundary range
+    for i, each in enumerate(nsents):
+        try:
+            sent_start_index = each[0][1][0]
+            sent_end_index = each[-1][1][1]
+            sent_bound_range[i] = (sent_start_index, sent_end_index)
+        except Exception as ex:
+            if i != len(nsents) - 1:
+                raise RuntimeError(f'The {i}th sentence is an empty sentence')
+
+    return nsents, sent_bound_range
+
+
+def __flat(data, to_str=False):
+    flatted = []
+
+    for each in data:
+        if isinstance(each, list) or isinstance(each, tuple):
+            for e in each:
+                flatted.append(e)
+        else:
+            flatted.append(each)
+
+    if to_str:
+        flatted = list(map(lambda x: str(x), flatted))
+
+    return flatted
+
+
+def BIOdata_to_file(file_name, sents, sep=" "):
+    # the data must be list of list
+    assert isinstance(sents, list), "the data object must be list and generated from generate_BIO()."
+    full_text = []
+    with open(file_name, "w") as fw:
+        # 'anticoagulant', (1000, 1013), (976, 989), 'B-Drug'
+        for sent in sents:
+            full_sent = []
+            for word in sent:
+                word = __flat(word, to_str=True)
+                full_sent.append(sep.join(word))
+                # word.append("\n")
+            full_text.append("\n".join(full_sent))
+        fw.write("\n\n".join(full_text) + "\n\n")
+
+
+def load_mapping_file(mapping_file, sep=" "):
+    with open(mapping_file, "r") as f:
```

Comment:
Recently I noticed a tool called: [refur](https://github.com/dosisod/refurb)
It recommends to: Use `y = Path(x).read_text()` instead of `with open(x, ...) as f: y = f.read()`
Anyway, just something I have noticed before. And there is no change needed.

Reply:
I will just keep as it is since both work fine